### DISCUSSION TITLE
Focus and unfocus window shows strange status message

### DISCUSSION THREAD

**Discussion by lemieuxl** [saulpw/visidata#2247] [General]
Hello all!

I need help with a strange message in the status bar of visidata which has been messing with my head for a while now.

When I start visidata (v3.0.1, python 3.12 / 3.11, fedora 39), I get the following message in the status bar: *no command for "kxOUShift+T"*. Visidata works as expected, but I cannot see the results of a one-time aggregator since only this message appears. It's been like this for a while now.

I tried removing my configuration file (`~/.visidatarc`), without any effects.

I don't think it's a bug, since visidata works as expected. I was wondering if any of you had this issue in the past.

Thanks!

**Comment by anjakefala** [comment]
@lemieuxl Could you tell us your terminal, `$TERM`, and Operating System?

**Reply by lemieuxl** [reply]
@anjakefala 
The message appears on gnome-console (45.0), gnome-terminal (3.50.1) and terminator (2.1.3). They all use `xterm-256color`. I'm using Fedora 39 (but the message was present on Fedora 38).

I also noticed that the message changes over time (when the window is in focus or not). When visidata first launches (and the window is in focus), I see `kxIShift+N`. When the terminal is out of focus (I click on a different window), I see `kxOUShift+T`. Finally, when I go back to the terminal, I see `kxIShift+N`.

Could visidata be registering something from wayland? The message appears at the same place where `LeftBtnUp`, `scrollUp`, etc. do appear.

**Comment by midichef** [comment]
Visidata is reacting to [terminal events for change of focus](https://stackoverflow.com/questions/73042942/python-curses-how-to-detect-focus/73043080#73043080).

You can try to [turn off the focus in/out events at the terminal level](https://superuser.com/questions/931873/o-and-i-appearing-on-iterm2-when-focus-lost/979294#979294). Try running `printf "\e[?1004l"` in your terminal before you run visidata.

For my system, gnome-terminal in Ubuntu, I can confirm that `printf "\e[?1004h"` turns on focus events that Visidata reacts to, and `printf "\e[?1004l"` turns them off. But instead of `kxIN` and `kxOUT`, I see `^[[I` and `^[[O`.

If you instead want to try to address the problem inside visidata, you can try this in `.visidatarc` as a quick solution:
```
Sheet.addCommand('kxIN', '', '', 'do nothing for focus-in event')
Sheet.addCommand('kxOUT', '', '', 'do nothing for focus-out events')
```

**Reply by saulpw** [reply]
Good find, @midichef!  (Or did you already know that bit of terminal trivia?)

A couple of points discovered during my shaving of this yak:

- we should use `no-op` instead of an empty string--makes it easier to search for.
- Since your system (and mine) doesn't parse this escape sequence, it's received as two separate keystrokes.  I've now added `Alt+[` as a prefix so unparsed ANSI escape sequences can trigger a command like this.
- To support the above, the `allPrefixes` mechanism has changed a bit, so that it has to contain all allowable prefix combinations.  This makes more sense anyway.
- All of this won't actually solve @lemieuxl 's problem because `no-op` is still a command and so the status panel will reset.

@lemieuxl did you know that `Ctrl+P` will bring up a sheet of past status messages?

I could also see us making a `z Ctrl+P` (`show-status-recent`) to repopulate the sidebar with the most recent status message(s).

**Reply by lemieuxl** [reply]
Thank you, both of you.

I confirm that the `printf` and `addCommand` workarounds did not work for me. And no, I did not know `Ctrl+P` showed a list of previous messages. At least now, I can see the aggregator result (though I could see some of them using `I`).

**Reply by saulpw** [reply]
@lemieuxl Are you saying that running `printf "\e[?1004l"` in gnome-terminal before running vd doesn't make the issue go away for you?  If so, I have no idea what's going on in your terminal or what could be done about it.  Please confirm.

**Reply by lemieuxl** [reply]
I confirm that running `printf` before `vd` has no effect (I still see the message).

```bash
printf "\e[?1004l" ; vd dummy.csv
```

**Reply by midichef** [reply]
@lemieuxl 
I reproduced this problem on a Fedora 39 machine. For now, try changing your shell's `$TERM` environment variable to something other than `xterm-256color`. Some choices that worked for me are:
`TERM=gnome-256color vd` and
`TERM=tmux-256color vd`
If those two have problems, you can try others listed when you run `toe -a`.

**Reply by midichef** [reply]
@saulpw 
This is all new to me, I didn't know about these focus events till now!

I can reproduce the problem, in an ssh session to a Fedora machine from gnome-terminal on Ubuntu. What I see is:

The `kxIN` event is sent to visidata when the terminal gets a genuine focus-in event, and the `kxOUT` for focus-out events. But oddly, the `kxIN` is also seen when visidata starts. (lemieuxl saw `kxOUT`-related messages, but I see messages for `kxIN`). And `kxIN` is seen after any line of input is read (for example, after `/` reads a regex or `z+` reads an aggregator). I'm guessing that something visidata does with curses is causing this event to be generated.

`Sheet.addCommand('kxIShift+N'` was able to add a command for the event, but `Sheet.addCommand('kxIN'` was not able to. Then, as you said, creating a command did not stop the problem of the `no-op` command overwriting the output that lemiuxl wanted to see.

`printf "\e[?1004l"` did nothing to stop these events in visidata. I believe that the 1004 mode is already off before visidata runs, because `printf "\e[?1004h"` caused new behavior in the shell (not in visidata). It made the terminal ring a bell when focus went in/out of the window, and `printf "\e[?1004l"` then restored the default shell behavior, of no bell.

[A thread from a few months ago on the vim repo](https://github.com/vim/vim/pull/13440) is the best description I can find so far for some of the recent changes around focus events and xterm. And there's [some recent advice for the developer of nano](https://lists.gnu.org/archive/html/bug-ncurses/2023-10/msg00087.html) on how to handle the focus events.

**Reply by lemieuxl** [reply]
@midichef
Changing `$TERM` to `gnome-256color` or `tmux-256color` fixes the issue. Now that I'm thinking about it, I did notice the status bar was fine when using *tmux*, but I don't use it that often locally (on my F39 machine).
